ompi/mca/coll/han/coll_han_gatherv.c

```diff
@@ -303,65 +314,49 @@ int mca_coll_han_gatherv_intra(const void *sbuf, int scount, struct ompi_datatyp
 
     /* #################### Node leaders ########################### */
 
-    uint64_t *low_data_size = NULL;
     char *tmp_buf = NULL;
-    ompi_datatype_t *temptype = MPI_BYTE;
 
     /* Allocate a temporary array to gather the data size, i.e. data type size x count,
      * in bytes from local peers */
-    low_data_size = malloc(low_size * sizeof(uint64_t));
-    if (!low_data_size) {
+    low_rcounts = malloc(low_size * sizeof(size_t));
+    if (!low_rcounts) {
         err = OMPI_ERR_OUT_OF_RESOURCE;
         goto node_leader_out;
     }
 
     /* Low Gather -  Gather local peers' send data sizes */
-    low_comm->c_coll->coll_gather((const void *) &send_size, 1, MPI_UINT64_T,
-                                  (void *) low_data_size, 1, MPI_UINT64_T, root_low_rank, low_comm,
+    low_comm->c_coll->coll_gather((const void *) &send_size, sizeof(size_t), MPI_BYTE,
+                                  (void *) low_rcounts, sizeof(size_t), MPI_BYTE, root_low_rank, low_comm,
                                   low_comm->c_coll->coll_gather_module);
 
-    /* Determine if we need to create a custom datatype instead of MPI_BYTE,
-     * to avoid count(type int) overflow
-     * TODO: Remove this logic once we adopt large-count, i.e. count will become 64-bit.
-     */
-    int total_up_scount = 0;
-    size_t rsize = 0, datatype_size = 1, max_data_size = 0;
-    for (int i = 0; i < low_size; ++i) {
-        rsize += (size_t) low_data_size[i];
-        max_data_size = (size_t) low_data_size[i] > max_data_size ? (size_t) low_data_size[i]
-                                                                  : max_data_size;
-    }
-
-    if (max_data_size > (size_t) INT_MAX) {
-        datatype_size = coll_han_utils_gcd(low_data_size, low_size);
-    }
-
-    low_rcounts = malloc(low_size * sizeof(int));
-    low_displs = malloc(low_size * sizeof(int));
-    tmp_buf = (char *) malloc(rsize); /* tmp_buf is still valid if rsize is 0 */
-    if (!tmp_buf || !low_rcounts || !low_displs) {
+    low_displs = malloc(low_size * sizeof(ptrdiff_t));
+    if (!low_displs) {
         err = OMPI_ERR_OUT_OF_RESOURCE;
         goto node_leader_out;
     }
 
+    size_t total_rsize = 0;
     for (int i = 0; i < low_size; ++i) {
-        low_rcounts[i] = (int) ((size_t) low_data_size[i] / datatype_size);
         low_displs[i] = i > 0 ? low_displs[i - 1] + low_rcounts[i - 1] : 0;
-        total_up_scount += low_rcounts[i];
+        total_rsize += low_rcounts[i];
     }
 
-    if (1 < datatype_size) {
-        coll_han_utils_create_contiguous_datatype(datatype_size, MPI_BYTE, &temptype);
-        ompi_datatype_commit(&temptype);
+    tmp_buf = (char *) malloc(total_rsize); /* tmp_buf is still valid if total_rsize is 0 */
+    if (!tmp_buf) {
+        err = OMPI_ERR_OUT_OF_RESOURCE;
+        goto node_leader_out;
```

Comment:
The behavior of `malloc` on zero-size allocations is implementation defined, so you cannot rely on getting a non-NULL pointer:

> If the size of the space requested is 0, the behavior is implementation-defined: either a null pointer shall be returned, or the behavior shall be as if the size were some non-zero value, except that the behavior is undefined if the returned pointer is used to access an object.

https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/functions/malloc.html

Reply:
Should this check if total_rsize is zero and set `err` if so? This code was copied from the original, so setting an error there might change the behavior.

Reply:
We should probably audit the code at some point to catch these. Let's leave this in for now.